rfcs/20200424-sparse-domain-isolation.md

```diff
@@ -0,0 +1,649 @@
+
+```python
+@tf_export("dynamic_embedding.embedding_lookup")
+def embedding_lookup(params,
+                     ids,
+                     name='embedding_lookup',
+                     max_norm=None):
+  """Provides a dynamic version of embedding_lookup
+    similar to tf.nn.embedding_lookup.
+
+  Ids are flattened to a 1d tensor before being passed to embedding_lookup
+  then, they are unflattend to match the original ids shape plus an extra
+  leading dimension of the size of the embeddings.
+
+  Args:
+    params: A dynamic_embedding.Variable instance.
+    ids: a tensor with any shape as same dtype of params.key_dtype.
+    name: A name for the operation (optional).
+    max_norm: If not `None`, each embedding is clipped if its l2-norm is larger
+      than this value.
+  Returns:
+    A `resource_variable_ops.TrainableWrap` which hold the handls of `params` and `ids`.
+  """
+  assert isinstance(params, Variable), "params should be a Variable instance."
+  assert params.key_dtype == ids.dtype, \
+    "params.value_dtype should be same with ids.dtype: {} vs. {}". \
+      format(params.key_dtype, ids.dtype)
+  vals = None
+  warp_vals = None
+  with ops.name_scope(name):
+    init = constant_op.constant(0.0, shape=(1,))
+    warp_vals = TrainableWrap(params,
+                              ids,
+                              initial_value=vals,
+                              dtype=params.value_dtype,
+                              trainable=params.trainable)
+    if max_norm is not None:
+      warp_vals = _clip(params.lookup(ids), ids, max_norm)
+  return warp_vals
+
+###
+
+@tf_export("dynamic_embedding.embedding_lookup_sparse")
+def embedding_lookup_sparse(params,
+                            sp_ids,
+                            sp_weights,
+                            name=None,
+                            combiner=None,
+                            max_norm=None):
+  """Provides a dynamic version of embedding_lookup_sparse
+    wich is similar to tf.nn.embedding_lookup_sparse in function.
+
+  Args:    
+     All args ars similar to `tf.nn.embedding_lookup_sparse`.
+  Returns:
+    A `resource_variable_ops.TrainableWrap` which hold the handles of `params` and `ids`.
+  """
+  pass
+```
+
+### Trainable Wrap
+
+In the early scheme, the `dynamic_embedding.Variable` will directly be trained by optimizers and **we had to extand all optimizers one-by-one**.
+
+Now we propose a new scheme which no longer require to extand all optimizers one by one: 
+We design a warp class `resource_variable_ops.TrainableWrap` which inherts from `resource_variable_ops.ResourceVariable`, and responsible for:
+
+- Maintain the relationship between `params` and `ids` created by `embedding_lookup{, _sparse}`.
+- Look up the newest values from `dynamic_embedding.Variable` and update them to memory held by `TrainableWrap->tensor()` before each iteration start.
+
+
+#### python/ops/resource_variable_ops.py:
+
+```python
+
+class TrainableWrap(ResourceVariable):
```

Comment:
My understanding of `TrainableWrap` is a container to hold (dynamic_embedding, ids) but pretends to be a tf.Variable. I am wondering whether it has to a `ResourceVariable`?

Reply:
> My understanding of `TrainableWrap` is a container to hold (dynamic_embedding, ids) but pretends to be a tf.Variable. I am wondering whether it has to a `ResourceVariable`?

For all of optimizers can train ResourceVariable through `resource_apply{,_sparse}` Or you has to extend the them one-by-one.